lib/ansible/modules/windows/win_zip.ps1

```diff
@@ -0,0 +1,253 @@
+#!powershell
+
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+#Requires -Module Ansible.ModuleUtils.Legacy
+#Requires -Version 5.0
+
+Set-StrictMode -Version 2.0
+
+$ErrorActionPreference = "Stop"
+#TODO/Future Enhancements
+#-------------------------
+# - support recurse or just content for dirs
+# - support multiple source files and dirs and globs
+# - "force" option to overwrite existing zip file
+# - Provide an 'update' parameter to add or remove files from the zip
+# - Support other archive types - $pcx_extensions = @('.bz2', '.gz', '.msu', '.tar', '.zip')
+
+# Notes
+# -----
+# The operations should be atomic - the zip shouldn't be created if anything fails
+
+# Initialize it here as it's used for the failifempty
+$result = @{
+    changed = $false
+    source_deleted = $false
+}
+
+$params = Parse-Args $args -supports_check_mode $true
+$check_mode = Get-AnsibleParam -obj $params -name "_ansible_check_mode" -type "bool" -default $false
+
+$src = Get-AnsibleParam -obj $params -name "src" -type "path" -failifempty $true -resultobj $result -aliases path
+$dest = Get-AnsibleParam -obj $params -name "dest" -type "path" -failifempty $true -resultobj $result -aliases zip_file
+$delete_src = Get-AnsibleParam -obj $params -name "delete_src" -type "bool" -default $false -aliases rm
+$creates = Get-AnsibleParam -obj $params -name "creates" -type "path"
+
+# Fixes a fail error message (when the task actually succeeds) for a
+# "Convert-ToJson: The converted JSON string is in bad format"
+# This happens when JSON is parsing a string that ends with a "\",
+# which is possible when specifying a directory to download to.
+# This catches that possible error, before assigning the JSON $result
+$result.dest = $dest -replace '\\$',''
+$result.src = $src -replace '\\$',''
+
+if ($creates) {
+    $result.creates = $creates
+}
+
+function Remove-BrokenZip([string]$dest) {
+    #Removes the broken zip
+    Remove-Item -LiteralPath $dest -Force -ErrorAction SilentlyContinue
+}
+
+function Remove-Source([string]$src) {
+    try {
+        if (-not $check_mode) {
+            if (Test-Path -LiteralPath $src) {
+                # -recurse won't cause issues for files
+                Remove-Item -LiteralPath $src -Recurse -Force -ErrorAction Stop
+            }
+        }
+        #$result.changed = $true
+        $result.source_deleted = $true
+    }
+    catch {
+        Remove-BrokenZip $dest
```

Comment:
Why would this now work, if you are catching an exception from `Remove-Item` then just subsequently calling it again it should also fail to remove the file.

Really you should just change the caller of `Remove-Source` to just do `Remove-Item -LiteralPath $src -Recurse -Force -WhatIf:$check_mode`. Will do the exact same thing and save you ~15 lines of code as well as making it clear what you are doing in that step instead of having something try and figure out what `Remove-Source` actually does.